pallets/dapp-staking-migration/src/lib.rs

```diff
@@ -0,0 +1,692 @@
+//! the old storage should be cleaned up using `RemovePallet` type.
+//!
+
+pub use pallet::*;
+
+use frame_support::{
+    dispatch::PostDispatchInfo,
+    log,
+    pallet_prelude::*,
+    traits::{Get, LockableCurrency, ReservableCurrency},
+};
+
+use frame_system::{pallet_prelude::*, RawOrigin};
+use parity_scale_codec::{Decode, Encode};
+use sp_io::{hashing::twox_128, storage::clear_prefix, KillStorageResult};
+use sp_runtime::{
+    traits::{TrailingZeroInput, UniqueSaturatedInto},
+    Saturating,
+};
+
+use pallet_dapps_staking::{Ledger as OldLedger, RegisteredDapps as OldRegisteredDapps};
+
+#[cfg(feature = "try-runtime")]
+use astar_primitives::Balance;
+#[cfg(feature = "try-runtime")]
+use sp_std::vec::Vec;
+
+pub use crate::pallet::DappStakingMigrationHandler;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+pub mod weights;
+pub use weights::WeightInfo;
+
+const LOG_TARGET: &str = "dapp-staking-migration";
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::config]
+    pub trait Config:
+        // Tight coupling, but it's fine since pallet is supposed to be just temporary and will be removed after migration.
+        frame_system::Config + pallet_dapp_staking_v3::Config + pallet_dapps_staking::Config
+    {
+        /// The overarching event type.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        /// Weight info for various calls & operations in the pallet.
+        type WeightInfo: WeightInfo;
+    }
+
+    /// Used to store the current migration state.
+    #[pallet::storage]
+    pub type MigrationStateStorage<T: Config> = StorageValue<_, MigrationState, ValueQuery>;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(crate) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Number of entries migrated from v2 over to v3
+        EntriesMigrated(u32),
+        /// Number of entries deleted from v2
+        EntriesDeleted(u32),
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Attempt to execute migration steps, consuming up to the specified amount of weight.
+        /// If no weight is specified, max allowed weight is used.
+        ///
+        /// Regardless of the specified weight limit, it will be clamped between the minimum & maximum allowed values.
+        /// This means that even if user specifies `Weight::zero()` as the limit,
+        /// the call will be charged & executed using the minimum allowed weight.
+        #[pallet::call_index(0)]
+        #[pallet::weight({
+            Pallet::<T>::clamp_call_weight(*weight_limit)
+        })]
+        pub fn migrate(
+            origin: OriginFor<T>,
+            weight_limit: Option<Weight>,
+        ) -> DispatchResultWithPostInfo {
+            ensure_signed(origin)?;
```

Comment:
We could, but for this I don't see the real need, tbh.

Migration needs to be done ASAP after we upgrade.

Do you have some specific scenario in mind though?

Reply:
Because the migration can be done in multiple steps with specific `weight_limit`s, I was worrying about the storage consistency. There could be such risk if there are activities in the old dApp staking pallet before all the migration steps executed.

But I double checked the code and noticed comment in line `527` indicates the old pallet would be put into maintenance mode. If it's in maintenance mode before the upgrade, there shouldn't be any risk.

Reply:
That's right, once upgrade is done, the old pallet becomes permanently inaccessible.
Not even root-privilege calls will be able to remove maintenance mode.

Reply:
I decided to add an additional type to (disable) enable this feature, since storage cleanup would eventually remove the maintenance storage setting from the old pallet.